src/react/4-props.md

```diff
@@ -0,0 +1,223 @@
+
+```js
+function addOneAndSeven(){
+    return 1 + 7
+}
+```
+
+This is a perfectly good function, but notice that it always adds the same two numbers, in other words, it's only usable in one very specific case.
+
+We can make this function more re-usable by allowing arguments to be passed into it:
+
+
+```js
+function addAndNumbers(num1, num2){
+    return num1 + num2
+}
+```
+
+The above function is now infinitely more re-usable because the user can specify the numbers, instead of just having them be hardcoded in.
+
+This same logic also applies to React components. 
+
+### Basic Props
+
+Recall that React components are essentially just functions (or classes) which return some JSX. And like normal functions, we run into the same re-usabiity problems. 
+
+
+```jsx
+function ErrorMessage() {
+  return <div className="error">There is an error</div>;
+}
+```
+
+Take a look at the `ErrorMessage` component above. It returns a div with the text `"There was an error"`. This is fine, but suppose we wanted to display a more specific message. In this case it's not possible because the error message is hard-coded in.
+
+We can solve this by modifying the `ErrorMessage` component so that it accepts props. 
+
+```jsx
+function ErrorMessage(props){
+  return (
+    <div className="error">{props.text}</div>
+  )
+}
+```
+
+In react, all functional components (and classes) receive a `props` object as their first argument (w/ classes this is the first argument in the constructor).
+
+The props object contains any values which are passed into the component when it's rendered.
+
+```jsx
+<ErrorMessage text="The username or password is invalid" />
+<ErrorMessage text="Error connecting to internet" />
+```
+
+Whenever the `ErrorMessage` component is rendered, we can now pass it a `text` prop which it will use as the displayed message.
+
+Now, instead of having an unchangeable, hard-coded message, the component is more flexible and can be re-used in any scenario where it's necessary to display an error message to the user.
+
+
+### Callback Props
+
+Just like normal function arguments, React components can accept props of any data type (even other components).
+
+One of the most useful types of props are callbacks. Callback props allow the programmer to specify what a component will do when an action occurs inside of it.
+
+```jsx
+function MyButton() {
+  return (
+    <button onClick={() => console.log('clicked')}>
+        click me
+    </button>
+  )
+}
+```
+
+The button above runs a simple function whenever the user clicks it. But once again, we're running into a re-usability issue.
+
+Suppose we wanted to perform a custom action when the button is clicked instead of the hard-coded `console.log('clicked')`.
+
+Here we would use a callback prop. We'll actually pass the component a callback function as one of it's props, which it will run whenever the button is clicked.
+
+```jsx
+function MyButton({handleClick}) {  // here we're destructuring the props object
```

Comment:
We should both mention and enforce the standard of starting all event props with `on`. `handle` is for the function you're passing in if it needs independent creation, but the prop should be `on`.